openvdb/tools/Count.h

```diff
@@ -0,0 +1,220 @@
+//////////////////////////////////////////////////////////////////////////
+//
+// TM & (c) Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved. This software is distributed under the
+// Mozilla Public License 2.0 ( http://www.mozilla.org/MPL/2.0/ )
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file Count.h
+///
+/// @author Dan Bailey
+///
+/// @brief Counting tools.
+
+#ifndef OPENVDB_TOOLS_COUNT_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_COUNT_HAS_BEEN_INCLUDED
+
+#include <openvdb/tree/NodeManager.h>
+
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+/// @brief Return a vector with node counts. The number of node of type
+/// NodeType is given as element NodeType::LEVEL in the return vector.
+/// Thus, the size if this vector corresponds to the height (or depth)
+/// of this tree.
+///
+/// @param tree       the tree to be counted
+/// @param threaded   enable or disable threading (threading is enabled by default)
+template <typename TreeT>
+std::vector<Index>
+nodeCount(const TreeT& tree, bool threaded = true);
+
```

Comment:
There's not a lot in it, with the large dataset Ken was testing with on my 32 logical core machine:

* Without threading - 7.9 ms
* With threading - 1.9 ms

I'm only threading over the root node hash table so there's still potentially a fair amount of traversal to do per thread and this method is a key component of the NodeManager which could be built quite frequently.

That being said it does complicate the code slightly and there's (currently) no logic to disable threading for smaller datasets so the overhead probably dominates in those cases. As a quick test, here's the same data set with all but one of the largest root node elements removed:

* Without threading - 0.32 ms
* With threading - 0.57 ms

And with a much smaller dataset:

* Without threading - 0.023 ms
* With threading - 0.040 ms

I frequently see logic in Houdini code to decide between whether to use threading or not (point count < 1024, etc). I'm curious to hear your thoughts on why you think that might be worthwhile? It is of course the fastest cases where the scheduling overhead dominates, so typically <1ms, do those really add up?

Reply:
Well, the problem is you never know if you are in the inner loop of another function or not :>

The most egregious is something like dot product of a vector.  Your vector could be 10 elements or 1 billion.  And it is not unusual to want to do 100,000 dot products on 100 element vectors, so this can add up.

Our 1024 is usually a wild guess...  I instrumented and set the vector stuff to 5000 many years ago, I think a recent test suggested more like 50,000 was a better cut off.  But some cut off I feel must always be there, and ideally it is done by the algorithm, not the caller, as the caller won't know which to use.  To address the debugging issue, our THREAED_ macros generate a NoThread version of every function.  And UTparallelFor has a drop in replacement of UTserialFor.

What can also be worse is that your 2x slower might be using more cores to be 2x slower, so be consuming not 2x, but 64x the core-cycles a single threaded algorithm would.  So now anything running on the system is impacted.

That said, I stand corrected about threading ever being necessary for this :>

Reply:
Some good points, thanks. One of the challenges with these simple algorithms is that there's a reasonable overhead in inspecting the data structure to figure out whether threading is worth doing or not, which may eradicate all or most of the benefit. In the case of nodeCount(), it would probably need to be something as trivial as thread when root node elements > 1.

I've been thinking about whether we should add a simple VDB abstraction around TBB's parallel_for/parallel_reduce calls for the same reason.